### ISSUE TITLE
Support Messaging Communication for Actors

### ISSUE THREAD

**Issue by amanbha** [dapr/dapr#724]
## Describe the proposal
Currently Actor framework in Dapr supports Request Response pattern. While this request response pattern is helpful in many scenarios it can cause issues in scenarios when service is busy or when client dies while waiting for response, if methods take longer to execute, it could affect the whole call chain upstream.
In addition to Request Response, also provide an option of communicating with Actors ny sending a message to an actor, so that client doesnt wait for service reply, service pulls work from queue when needed. At higher level this can be seen as a variation of trigerring of Actors.

cc @mkosieradzki

**Comment by davidfowl** [comment]
What does that mean from an API POV? What does the calling pattern look like?

**Comment by yaron2** [comment]
> What does that mean from an API point of view? What does the calling pattern look like?

It would look the same, only the response will return as soon as the message has been put into the actor queue.

**Comment by davidfowl** [comment]
So still RPC like but only for methods that don’t return a value?

**Comment by yaron2** [comment]
> So still RPC like but only for methods that don’t return a value?

Yes. On the flip side, we could make a more publish/subscribe type of API, but personally that would take some of the typed method experience provided by actor sdks.

Wdyt?

**Comment by davidfowl** [comment]
There are a couple of random thoughts I had:
- How is back pressure communicated to the caller? If the actor queue gets flooded, will the caller stop? If yes, how is that managed?
- Is there value in the caller knowing whether the method executed on the other actor? If yes then you’d want something at the callsite to distinguish between the call executed and the message has been enqueued.
- Who is responsible for deciding fire and forget semantics? The caller or the actor itself? 

> On the flip side, we could make a more publish/subscribe type of API, but personally that would take some of the typed method experience provided by actor sdks.

You could imagine a pattern like:

actor.Queue.Method()
actor.Method()

The query property is how you dispatch to the queue and the direct call is rpc, wait for a response.

**Comment by wanton7** [comment]
Our company used Orleans for one proof of concept app. In Orleans state is usually loaded when actor starts and callers will receive return values very fast because state is cached inside actor. In Orleans you can also mark state reading methods to be reentrant so your writes don't block readers, this is possible because Orleans uses custom task scheduler.

I've tried Dapr and way it works you can't even return same state to all waiting calls. Example if you have UserActor with method like GetInfo that would return Id, Email and Name and if that method would be called 100 times simultaneous you would have to call state store to get state 100 times.

If communication would be message based maybe actor could know what messages are waiting and see if there are multiple messages that you could satisfy and respond with state you got from state store, maybe with configurable message box limit per actor. In GetInfo method's case you could get state from state store once and would use that single state for all 100 message responses.

**Comment by yaron2** [comment]
The state of the actor in Dapr is cached as well.
100 callls will result in 1 call to the actual state store.

/cc @amanbha

**Comment by wanton7** [comment]
Oh didn't know that, that's awesome! I read the docs at https://github.com/dapr/docs and I didn't see any mention of that, but it could be that I just missed it.

**Comment by yaron2** [comment]
Or it could be we didn't document it properly, actors doc are a little sparse.

But there's client side caching of state with actors.

**Comment by amanbha** [comment]
Yes, the dotnet sdk and java sdk has a cache built in for actor states.

**Comment by marxxxx** [comment]
When I have a chain of, say, 3 actors calling each other, is the first call blocking until the last method call executed completely? If this is the case, is there a way to get around this behavior in case I need the semantics of a queue?

**Comment by artursouza** [comment]
> When I have a chain of, say, 3 actors calling each other, is the first call blocking until the last method call executed completely? If this is the case, is there a way to get around this behavior in case I need the semantics of a queue? 

A given actor instance (ActorType + ActorId) can only handle one request at a time - other calls are blocked waiting. We have plans for two features: pubsub and reentrancy for actors. I don't have the issue url handy now since I am replying from my phone.

**Comment by ruslanguk** [comment]
Hi, 

Here is our scenario: 

We have a large number of “short” running tasks but with a lot of dependencies. On the diagram it shows that Task1 can’t perform its execution until other tasks are completed.
Task2 can’t complete its execution until Task4 and Task7 are completed and so on. The tree is pretty large and the Task1(actor) might wait (deactivated)  several hours until the rest of the actors have performed their job.
This example is simplified but the number of actors is very large and the number of dependencies is even larger (10x).  

This can be easily solved in AKKA with a Fire-And-Forget messaging between actors (but not distributed across many nodes of course and we are looking for scalable solution).  
 
![Tasks Dependency](https://user-images.githubusercontent.com/51964403/124633550-c1d0aa00-de5b-11eb-9066-be6636fbe3e5.png)
